exercises/template/ex02_oo_basics/src/shape.hpp

```diff
@@ -0,0 +1,85 @@
+#ifndef SHAPE_HPP
+
+#define SHAPE_HPP
+
+#include <iostream>
+#include <math.h>
+#include <memory>
+#include <vector>
+
+using namespace std;
+
+class Shape {
+public:
+  virtual int sides() const = 0;
+  virtual double perimeter() const = 0;
+  virtual double area() const = 0;
+  virtual string type() const = 0;
+};
+
+ostream &operator<<(ostream &os, const Shape *shape);
+
+class Circle : public Shape {
+private:
+  double radius;
+
+public:
+  explicit Circle(double r) : radius(r) {}
+  int sides() const override;
+  double perimeter() const override;
+  double area() const override;
+  string type() const override;
+};
+
+class Square : public Shape {
+private:
+  double side;
+
+public:
+  explicit Square(double s) : side(s) {}
+  int sides() const override;
+  double perimeter() const override;
+  double area() const override;
+  string type() const override;
+};
+
+class Rectangle : public Shape {
+private:
+  double width;
+  double height;
+
+public:
+  explicit Rectangle(double w, double h) : width(w), height(h) {}
+  int sides() const override;
+  double perimeter() const override;
+  double area() const override;
+  string type() const override;
+};
+
+class Triangle : public Shape {
+private:
+  double base;
+  double height;
+
+public:
+  explicit Triangle(double b, double h) :base(b), height(h) {}
+  int sides() const override;
+  double perimeter() const override;
+  double area() const override;
+  string type() const override;
+};
+
+class ShapeSorter {
+private:
+  vector<unique_ptr<Shape>> contents;
+
+public:
+  explicit ShapeSorter(vector<unique_ptr<Shape>> &c) : contents(move(c)){};
```

Comment:
use `vector<unique_ptr<Shape>> &&c` also, do we need to use pointers?

Reply:
I've looked into trying to avoid using a pointer here, but I'm not sure how to do that while still allowing the vector to be heterogeneous for all of the subclasses of Shape.  The best alternate solutions that I came up with were doing either a union or a variant.  The variant had the obvious downside of requiring C++17.  Both solutions had the ugliness of needing to add in each child class as it is added.

I'm likely missing something, but it looks like we do need pointers here if multiple different types will go into the same vector.